draft-ietf-jsonpath-base.md

```diff
@@ -1411,82 +1410,101 @@ a type system is first introduced.
-* For it and any function expression it contains,
-each argument of the function matches the defined type of the argument
-according to one of the following rules:
-  * The argument is a function expression with defined result type
-    that is the same as, or a subtype of, the defined type of the argument.
-  * The argument is a literal primitive value and the defined type of the  argument is `Value` or any type of which `Value` is a subtype.
-  * The argument is a Singular Path and the defined type of the argument is `OptionalNodeType` or any type of which `OptionalNodeType` is a subtype.
-  * The argument is a `filter-path` (which includes Singular Paths) and the defined type of the argument is `OptionalNodesType`.
+A function expression is typed based on its return, and may only appear in expressions as follows:
+
+* If the function return is of type `ValueType`, it may only
+  appear as a `comparable` in a comparison.
+* If the function return is of type `BooleanType`, it may
+  only appear in a `test-expr`.
+* If the function return is of type `NodelistType`, it may
+  appear either as a `comparable` in a comparison or in
+  a `test-expr` as members of `NodelistType` can be converted
+  to either `ValueType` or `BooleanType`.
+
+#### A Note of Caution for Defining Functions and Authoring Paths
+
+While functions returning `NodelistType` can appear in both comparisons and test expressions,
+path authors must be aware of the conversions and the potential outcomes.
+Specifically, nodelists convert to values and booleans according to {{tbl-typeconv}}:
+
+| Function return | Converted Value | Converted Boolean |
+| :-: | :-: | :-: |
+| empty nodelist | `Nothing` | `BooleanFalse` |
+| single-node nodelist | the node's JSON value | `BooleanTrue` |
+| multiple-node nodelist | `Nothing` | `BooleanTrue` |
+{: #tbl-typeconv title="Conversion of nodelists to values and booleans"}
+
+For example, given a `NodelistType` function `myfunc()`, the following paths
+are both valid but can have different and unexpected results:
+
+| function return | `myfunc(@.a)` evaluation (as a test) | `myfunc(@.a)==42` evaluation (in a comparison) |
+| :-: | :- | :- |
+| empty nodelist | The nodelist is converted to `BooleanFalse`.<br>The expression result is false. | The empty nodelist is converted to `Nothing`.<br>The expression result is false. |
+| single-node nodelist| The nodelist is converted to `BooleanTrue`.<br>The expression result is true. | The nodelist is converted to the node's value.<br>The expression result is true. |
+| mutliple-node nodelist| The nodelist is converted to `BooleanTrue`.<br>The expression result is true. | The nodelist is converted to `Nothing`.<br>The expression result is false. |
+
+The expression outcome is different between the test expression and comparison
+in the multiple-node nodelist case.
+This is a result of the type system and conversions, and it cannot be resolved
+in a way that keeps the type system consistent.
```

Comment:
Perhaps removing conversions would resolve the issue. However, I don't think the issue is type system inconsistency. The type system seems consistent to me. It's just that the behaviour might be a little surprising for people who are not familiar with the spec.

Reply:
As stated before, the conversions are necessary, and "removing" them is just describing them without naming them.  If you don't think list 1496 and 1497 are needed, I can remove them, but I thought it'd be fair to offer some explanation as to why the type system is expressed the way it is.